scenarios/remote_signals/src/main.rs

```diff
@@ -0,0 +1,239 @@
+                    let msg = match msg {
+                        Signal::App { signal, .. } => {
+                            let msg: Vec<u8> = signal.into_inner().decode().unwrap();
+                            let msg: TimedMessage = ExternIO(msg).decode().unwrap();
+                            msg
+                        }
+                        _ => continue,
+                    };
+
+                    pending_set.lock().unwrap().remove(&msg.to_request());
+
+                    match msg {
+                        TimedMessage::TimedRequest { .. } => (),
+                        TimedMessage::TimedResponse { requested_at, .. } => {
+                            let dispatch_time_s = requested_at.as_micros() as f64 / 1_000_000.0;
+                            let receive_time_s = Timestamp::now().as_micros() as f64 / 1_000_000.0;
+
+                            reporter.add_custom(
+                                ReportMetric::new("remote_signal_round_trip")
+                                    .with_field("value", receive_time_s - dispatch_time_s),
+                            );
+                        }
+                    }
+                }
+            });
+
+            Ok(())
+        })?;
+
+    install_app(
+        ctx,
+        scenario_happ_path!("remote_signal"),
+        &"remote_signal".to_string(),
+    )?;
+    try_wait_for_min_peers(ctx, Duration::from_secs(120))?;
+
+    Ok(())
+}
+
+fn agent_behaviour(
+    ctx: &mut AgentContext<TryCPRunnerContext, TryCPAgentContext<ScenarioValues>>,
+) -> HookResult {
+    let client = ctx.get().trycp_client();
+
+    let agent_name = ctx.agent_name().to_string();
+    let app_port = ctx.get().app_port();
+    let cell_id = ctx.get().cell_id();
+    let next_remote_signal_peer = ctx.get_mut().scenario_values.remote_signal_peers.pop();
+    let signal_interval = ctx.get().scenario_values.signal_interval;
+    let response_timeout = ctx.get().scenario_values.response_timeout;
+    let pending_set = ctx.get().scenario_values.pending_set.clone();
+    let reporter = ctx.runner_context().reporter();
+
+    let new_peers = ctx
+        .runner_context()
+        .executor()
+        .execute_in_place(async move {
+            let now = Timestamp::now();
+
+            pending_set.lock().unwrap().retain(|msg| {
+                if now.as_millis() - msg.requested_at().as_millis()
+                    > response_timeout.as_millis() as i64
+                {
+                    reporter.add_custom(
+                        ReportMetric::new("remote_signal_timeout").with_field("value", 1),
+                    );
+                    false
+                } else {
+                    true
+                }
+            });
+
+            match next_remote_signal_peer {
+                None => {
+                    // No more agents available to signal, get a new list.
+                    // This is also the initial condition.
+                    let mut new_peer_list = client
```

Comment:
This is seeming like a common pattern now, maybe we should consider looking for a way to extract it. Not a task for this PR though!